panel_client/panel_client.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import locale
+import json
+import multiprocessing
+from Queue import Empty as EmptyQueueException
+import os
+import socket
+import struct
+import sys
+import time
+
+# Polling interval of the IO process
+MIN_LATENCY_MS = 10
+
+# characer that delimits socket messages
+DELIMITER_CHAR = '\r'
+
+
+class PanelStateBase:
+
+  def diff_states(old_state, new_state):
+    return {
+        k:v for k in new_state.iteritems()
+        if (k not in old_state) or (v != old_state[k])
+    }
+
+  def get_state_updates(self):
+    """Must implement this function for your panel.
+
+    Should return an iterable of control_id, state pairs for anything that has
+    changed since the last call.
+    """
+    raise NotImplementedError()
+
+  def get_controls(self):
+    """Must implement this function for your panel.
+
+    Should return the available controls.
+    """
+    raise NotImplementedError()
+
+  def display_message(self, message):
+    """Must implement this function for your panel.
+
+    Display a message from the server for the user."""
+
+
+def _make_update_message(update):
+  return {'message': 'set-state', 'data': {'id': update[0], 'state': update[1]}}
+
+
+def _make_announce_message(controls):
+  return {'message': 'announce', 'data': {'controls':controls}}
+
+
+def _panel_io_subprocess_main(panel_state_factory, action_queue, message_queue):
+  panel_state = panel_state_factory()
+  action_queue.put(_make_announce_message(panel_state.get_controls()))
```

Comment:
Btw panel classes should be able to re-announce if they detect that controls disconnect / reconnect (so that the game could continue without our intervention), so might want to make this line its own method.

(Dawn was saying she didn't know if her panel (at least) would have the capability to detect disconnection, but I think it would be a nice gesture and the server supports it.)

Reply:
Will add as TODO, and implement if the panels actually end up able to detect this.